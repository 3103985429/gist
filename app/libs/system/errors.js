let GlobalError = Error;

if (process.env.NODE_ENV === 'testing') {
    // Workaround for `new PublicError() instanceof PublicError` returning false
    GlobalError = class {
        constructor(message) {
            this.message = message;
            this.stack = null;
        }

        toString() {
            return `Error: ${this.message}`;
        }
    };
}

class ExtendableError extends GlobalError {
    constructor(message) {
        super(message);
        this.name = this.constructor.name;
        this.stack = (new GlobalError(message)).stack;
    }
}

class PublicError extends ExtendableError {
}

class UserError extends PublicError {
    constructor(message) {
        super(message);
        this.statusCode = 400;
    }
}

class RequestError extends ExtendableError {
    constructor(message) {
        super(message);
        this.statusCode = 400;
    }
}

class ForbiddenError extends ExtendableError {
    constructor(message) {
        super(message);
        this.statusCode = 403;
    }
}

module.exports = {
    PublicError,
    UserError,
    RequestError,
    ForbiddenError
};
