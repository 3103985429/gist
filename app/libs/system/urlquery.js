// 'use strict'; //?

/**
 * URL Query...
 * @module  urlquery
 */

const winston = require('winston');
const querystring = require('querystring');

module.exports = function (str, callback) {

    class MyError extends Error {}

    query = querystring.parse(str);

    // var source_key = req.params.source_key;
    // if (source_key == null || typeof source_key == 'undefined' || source_key == "") {
    // return next(err);
    // }
    //
    // return callback(new MyError('there was a problem'));
    winston.error(str);
    winston.error(query);

    return callback(null, query);
};
