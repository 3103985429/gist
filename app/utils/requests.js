const winston = require('winston');

const { PublicError } = require('../libs/system/errors');

module.exports = {
    handleError,
};

/**
 * Returns true when an error occurred.
 * @param {Error} err
 * @param {Object} res
 * @param {String=} message
 * @returns {Boolean}
 */
function handleError(err, res, message = 'An unexpected error happened during request.') {
    if (err) {
        const code = err.statusCode || 500;
        let responseMessage = message;

        if (err instanceof PublicError) {
            responseMessage = err.message;
        } else if (['development', 'testing'].includes(process.env.NODE_ENV)) {
            responseMessage = `Non-public error: ${err.message}`;
        }

        winston.error(message, err);
        res.status(code).json({ error: responseMessage });

        return true;
    }

    return false;
}
