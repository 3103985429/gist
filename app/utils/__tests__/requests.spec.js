import { handleError } from '../requests';
import { PublicError } from '../../libs/system/errors';


describe('Requests Utils', () => {
    describe('when #handleError is called', () => {
        let res;

        beforeEach(() => {
            res = {
                data: {},
                status(status) {
                    this.data.status = status;
                    return this;
                },
                json(data) {
                    this.data.data = data;
                    return this;
                },
            };
        });

        it('should return false when no error exists', () => {
            const result = handleError(null, {});

            expect(result).toBe(false);
        });

        it('should respond with public error', () => {
            const result = handleError(new PublicError('public error'), res);

            expect(result).toBe(true);
            expect(res.data).toMatchSnapshot();
        });

        it('should respond with server error', () => {
            const oldNodeEnv = process.env.NODE_ENV;
            process.env.NODE_ENV = 'production';
            const result = handleError(new Error('example error'), res);
            expect(result).toBe(true);

            expect(res.data).toMatchSnapshot();
            process.env.NODE_ENV = oldNodeEnv;
        });

        it('should respond with custom error', () => {
            const oldNodeEnv = process.env.NODE_ENV;
            process.env.NODE_ENV = 'production';
            const result = handleError(new Error('example error'), res, 'Custom response');

            expect(result).toBe(true);
            expect(res.data).toMatchSnapshot();
            process.env.NODE_ENV = oldNodeEnv;
        });
    });
});
