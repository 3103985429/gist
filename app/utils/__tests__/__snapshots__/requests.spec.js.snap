// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Requests Utils when #handleError is called should respond with custom error 1`] = `
Object {
  "data": Object {
    "error": "Custom response",
  },
  "status": 500,
}
`;

exports[`Requests Utils when #handleError is called should respond with public error 1`] = `
Object {
  "data": Object {
    "error": "public error",
  },
  "status": 500,
}
`;

exports[`Requests Utils when #handleError is called should respond with server error 1`] = `
Object {
  "data": Object {
    "error": "An unexpected error happened during request.",
  },
  "status": 500,
}
`;
